compiler/optimizing/induction_var_simplification.cc

```diff
@@ -0,0 +1,318 @@
+#include "induction_var_simplification.h"
+#include<tuple>
+
+namespace art{
+
+
+typedef std::tuple<HInstruction*,HConstant*, HConstant* > triple;
+//
+// public methods
+//
+InductionVarSimplification::InductionVarSimplification(HInductionVarAnalysis* induction_analysis)
+    : induction_analysis_(induction_analysis),
+      biv_increment_(nullptr) {
+  DCHECK(induction_analysis != nullptr);
+}
+
+/*Helper Functions*/
+static HConstant* GetNewConstant(HInstruction::InstructionKind kind, HConstant* cst1, HConstant* cst2) {
+   DataType::Type type = cst1->GetType();
+   HConstant* new_cst = nullptr;
+   HGraph* graph = cst1->GetBlock()->GetGraph();
+   if (DataType::IsIntegralType(type)) {
+     int64_t new_val = 0;
+     switch (kind) {
+      case HInstruction::kAdd :
+      new_val = Int64FromConstant(cst1) + Int64FromConstant(cst2);
+      break;
+      case HInstruction::kSub: 
+      new_val = Int64FromConstant(cst1) - Int64FromConstant(cst2);
+      break;
+      case HInstruction::kMul:
+       new_val = Int64FromConstant(cst1) * Int64FromConstant(cst2);
+       break;
+      default :
+       UNREACHABLE();
+     }
+      new_cst = graph->GetConstant(type, new_val);
+      return new_cst;
+    } else if (cst1->IsDoubleConstant()) {
+      double new_val =0.0;
+      switch (kind){
+        case HInstruction::kAdd:
+        new_val = cst1->AsDoubleConstant()->GetValue() + cst2->AsDoubleConstant()->GetValue();
+        break;
+        case HInstruction::kSub:
+        new_val = cst1->AsDoubleConstant()->GetValue() - (cst2->AsDoubleConstant()->GetValue());
+        break;
+        case HInstruction::kMul:
+        new_val = cst1->AsDoubleConstant()->GetValue() * cst2->AsDoubleConstant()->GetValue();
+        break;
+        default :
+          UNREACHABLE();
+      }
+      new_cst = graph->GetDoubleConstant(new_val);
+      return new_cst;
+    } else if (cst1->IsFloatConstant()) {
+      float new_val = 0.0;
+      switch (kind) {
+       case HInstruction::kAdd:
+        new_val = cst1->AsFloatConstant()->GetValue() + cst2->AsFloatConstant()->GetValue();
+        break;
+       case HInstruction::kSub:
+        new_val = cst1->AsFloatConstant()->GetValue() - cst2->AsFloatConstant()->GetValue();
+        break;
+       case HInstruction::kMul:
+        new_val = cst1->AsFloatConstant()->GetValue() * cst2->AsFloatConstant()->GetValue();
+        break;
+        default:
+          UNREACHABLE();
+      }
+      new_cst = graph->GetFloatConstant( new_val);
+      return new_cst;
+    }
+    return nullptr;
+}
+static HConstant* GetZeroConstant(HInstruction* input){
```

Comment:
is it possible to have GetConstofTypeAndValue (value, type)?

Reply:
Done